editor/modules/runner/game-engine.ts

```diff
@@ -0,0 +1,143 @@
+import {Injectable} from '@angular/core';
+
+import {CommandBuffer, ClearAll, FlipY} from '../../rendering/commands';
+import {SpriteProgram, TileProgram} from '../../rendering/shaders';
+import {SpriteBatchProgram} from '../../rendering/shaders';
+import {DefaultFBO} from '../../rendering/fbo';
+import {Camera} from '../../rendering/camera';
+import {SpriteHandle, SpriteBatchObject} from '../../rendering/sprite';
+
+import {WebGLSurface} from '../../components';
+
+import {Map} from '../../models/map';
+import {PhysicsEngine} from './physics-engine';
+import {GameInput} from './game-input';
+import {GameState} from './game-state';
+import {Player} from './player';
+import {LycanService} from './lycan.service';
+import {Hud} from './hud';
+
+
+@Injectable()
+export class GameEngine {
+
+    private surface: WebGLSurface;
+    private camera = new Camera();
+    private last_time: number;
+    private player_handle: SpriteHandle;
+    private sprites: SpriteBatchObject;
+
+    constructor(
+        private lycan: LycanService,
+        private input: GameInput,
+        private world: GameState,
+        private player: Player,
+        private physics: PhysicsEngine,
+    ) {}
+
+    init(surface: WebGLSurface) {
+        this.surface = surface;
+        this.surface.setKeyHandler(this.input);
+        this.surface.setMouseHandler(this.input);
+        this.surface.addViewportListener(this.camera);
+    }
+
+    cleanUp() {
+        this.surface.removeViewportListener(this.camera);
+        this.surface.setKeyHandler();
+        this.surface.setMouseHandler();
+    }
+
+    play(map: Map) {
+        let chipsets_pos: {[path: string]: number} = {};
+        let chipsets_path: string[] = [];
+        map.fillChipsetsInfo(chipsets_pos, chipsets_path);
+
+        let map_tiled = this.surface.createTilesRenderEl();
+        map_tiled.loadTileLayerObject(chipsets_path, (chipsets, builder) => {
+            let handle = builder.setWidth(map.width)
+                .setHeight(map.height)
+                .tileSize(map.tile_size);
+            for (let i = 0; i < map.layers.length; ++i) {
+                let layer = map.layers[i].raw.map(pl => {
+                    return {
+                        tiles_id: pl.tiles_id,
+                        chipset: chipsets[chipsets_pos[pl.chipset]]
+                    };
+                });
+                handle.addLayer(layer);
+            }
+            // this.camera.centerOn(handle.build());
```

Comment:
If you are leaving this line for a reason, you should add a sentence to explain why.